drivers/ipcc/ipcc_write.c

```diff
@@ -0,0 +1,241 @@
+ *   buflen - size of the buffer... buffer
+ *
+ * Returned Value:
+ *   Number of successfully written bytes into the IPCC memory or netagted
+ *   errno when no data could be written.
+ *
+ * Assumptions/Limitations:
+ *
+ ****************************************************************************/
+
+ssize_t ipcc_write(FAR struct file *filep, FAR const char *buffer,
+                         size_t buflen)
+{
+  FAR struct ipcc_driver_s *priv;
+  ssize_t nwritten;
+  int ret;
+
+  /* Get our private data structure */
+
+  DEBUGASSERT(filep != NULL && filep->f_inode != NULL);
+  priv = filep->f_inode->i_private;
+
+  /* Get exclusive access to driver */
+
+  if ((ret = nxsem_wait(&priv->exclsem)))
+    {
+      /* nxsem_wait() will return on signal, we did not start
+       * any transfer yet, so we can safely return with error
+       */
+
+      return ret;
+    }
+
+  for (nwritten = 0; ; )
+    {
+#ifdef CONFIG_IPCC_BUFFERED
+      /* Buffered write, if buffer is empty try to write directly to
+       * IPCC memory, else buffer data in circbuf - it will be written
+       * in interrupt handler when IPCC tx channel is free.
+       */
+
+      if (circbuf_used(&priv->ipcc->txbuf) == 0)
+        {
+          /* Write buffer is empty, we can try and write data directly
+           * to IPCC memory thus omitting copying to buffer.
+           */
+
+          nwritten += priv->ipcc->ops.write(priv->ipcc, buffer + nwritten,
+                                           buflen - nwritten);
+          if (nwritten == (ssize_t)buflen || nwritten < 0)
+            {
+              /* We've managed to write whole buffer to IPCC memory,
+               * there is nothing else for use to do
+               *   --or--
+               * lower half driver returned error during write,
+               *
+               * either way we return with nwritten which will either
+               * be number of bytes written or negated errno.
+               */
+
+              nxsem_post(&priv->exclsem);
+              return nwritten;
+            }
+        }
+
+      /* Either, there is already some data on the txbuffer, which
+       * means IPCC is occupied, or txbuffer is empty, but we could
+       * not write whole buffer to IPCC memory. In either case we
+       * copy what is left in data to buffer.
+       */
+
+      nwritten += circbuf_write(&priv->ipcc->txbuf, buffer + nwritten,
+                                (ssize_t)buflen - nwritten);
+
+      /* Notify lower half that new data on circ buffer is available */
+
+      priv->ipcc->ops.write_notify(priv->ipcc);
+
+      if (nwritten == (ssize_t)buflen)
+        {
+          /* All outstanding data has been copied to txbuffer, we're done */
+
+          nxsem_post(&priv->exclsem);
+          return nwritten;
+        }
+
+#else /* CONFIG_IPCC_BUFFERED */
+      /* Unbuffered write, write data directly to lower driver */
+
+      nwritten += priv->ipcc->ops.write(&priv->ipcc, buffer, buflen);
```

Comment:
how to handle the fail at the 2nd or 3rd loop?

Reply:
You don't. IPCC write() can't really fail. IPCC uses only RAM and some registers to send data. These things cannot really fail, and if they do, then it's a critical hardware failure that I don't think we should care about.

There is possibility that this will always return 0, but in that case second CPU simply does not consume data or never informs us that data has been read - it's a bug in IPCC implementation of second CPU. With that bug we can also hang in semaphore waiting for the interrupt.

If user expects such situations, O_NONBLOCK then should be used.

That said, that nwritten < 0 check is useless and I will remove it.